### ISSUE TITLE
Unhandled project rejection

### ISSUE THREAD

**Issue by gavinrobertszeco** [dwyl/aws-sdk-mock#53]
Hi there,

**TL;DR: Mock rejects on error even when a callback is used.**

When trying to mock an S3 `getObject` request to return an error message, Node is complaining about an unhandled promise rejection, despite all promises being handled accordingly and working correctly without the mock.

    function doSomething(param) {
      return new Promise(function(resolve, reject) {
        // Get the S3 Service.
        var s3 = new AWS.S3();
        // Create a queue of promises to work with.
        var queue = [];
        // Loop through each file.
        param.forEach(function(file) {
        // Create a promise for each file to download.
          queue.push(new Promise(function(_resolve, _reject) {
            // Get the file
            s3.getObject({
              Bucket: '...',
              Key: '...'
            }, function(err, response) {
              // If we have an error, siliently fail.
              if (err) {
                console.log(err);
                return _resolve(false);
              }
              // Resolve with the result.
              _resolve(response);
            });
          });
        });
      });
      // Attempt to download all files.
      Promise.all(queue).then(function(results) {
        // Failed results are false - we'll strip out falsey 
        // values, leaving only those that were successful.
        results = results.filter(el => el);
        // If we don't have anything to continue with, we'll hard fail.
        if (!results || results.length > 0) {
          return reject('No results');
        }
        resolve();
      }).catch(function() {
        reject('Something bad happened');
      });  
    }

And then running:

    doSomething(['...']).then(function() {
      throw new Error("Promise should be rejected!");
    }).catch(function(err) {
      expect(err).to.equal('No results');
    });

So... The above code, without a mock works as expected. I'm passing it an invalid S3 key, resulting in the correct `NoSuchKey` error being thrown:

    {
      code: 'NoSuchKey',
      message: 'The specified key does not exist.'
    }

However when mocking that error:

    AWS.mock('S3', 'getObject', function(config, callback) {
      callback({
        code: 'NoSuchKey',
        message: 'The specified key does not exist.'
      });
    });

Causes the following to appear in my console, despite all tests passing correctly:

    (node:4503) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): [object Object]

Listening for `unhandledRejection` results in the error object I'm passing from the mock.

    process.on('unhandledRejection', (reason, promise) => {
        console.log(reason);
    });

Looking at the code for `aws-sdk-mock` - despite me using a callback rather than promises, it's rejecting a promise instead of returning the error via the callback as per the actual AWS SDK.

Any thoughts? I didn't realise that AWS supported promises, so I can update my code but for those using callbacks, I imagine this is a bug?

**Comment by gavinrobertszeco** [comment]
FYI - I've updated my code to use promises however for those still waiting to use callbacks, should this be looked into?

**Comment by jcalderonzumba** [comment]
@gavinrobertszeco I do think the same, someone should look upon it, this is happening also in my tests.

**Comment by mrowles** [comment]
Would love a hand with this

**Comment by bchociej** [comment]
Hi,

I believe that this is happening because a rejected Promise is being created on this line even if the caller never uses the `.promise()` form of the request https://github.com/dwyl/aws-sdk-mock/blob/master/index.js#L115

I think this can be fixed by waiting to do `makeRejected()` or `makeResolved()` until someone actually calls `request.promise()`: https://github.com/dwyl/aws-sdk-mock/blob/master/index.js#L130

I hope this is helpful. It's very late where I am or I'd take a swing at fixing it, but please if you would like me too, just @ me and I'll give it a go when I've got more energy.

Love this library! Thank you!
